File: packages/android/src/main/kotlin/com/generativeui/dsl/snapshot/RenderSnapshot.kt
package com.generativeui.dsl.snapshot

import com.generativeui.dsl.model.UINode

object RenderSnapshot {

    fun generate(node: UINode, indent: Int = 0): String {
        val sb = StringBuilder()
        appendNode(sb, node, indent, inRow = false)
        return sb.toString().trimEnd()
    }

    private fun appendNode(sb: StringBuilder, node: UINode, indent: Int, inRow: Boolean) {
        val prefix = "  ".repeat(indent)
        when (node) {
            is UINode.VStack -> {
                sb.appendLine("${prefix}VStack(spacing=${node.spacing ?: 8.0}, align=${node.alignment ?: "center"}) [children=${node.children.size}]")
                node.children.forEach { child ->
                    appendNode(sb, child, indent + 1, inRow = false)
                }
            }
            is UINode.HStack -> {
                sb.appendLine("${prefix}HStack(spacing=${node.spacing ?: 8.0}, align=${node.alignment ?: "center"}, fillsWidth=true) [children=${node.children.size}]")
                node.children.forEach { child ->
                    appendNode(sb, child, indent + 1, inRow = true)
                }
            }
            is UINode.ZStack -> {
                sb.appendLine("${prefix}ZStack(align=${node.alignment ?: "center"}) [children=${node.children.size}]")
                node.children.forEach { child ->
                    appendNode(sb, child, indent + 1, inRow = false)
                }
            }
            is UINode.Text -> {
                sb.appendLine("${prefix}Text(content=\"${node.content}\", style=${node.style ?: "body"}, color=${node.color ?: "default"}, weight=${node.weight ?: "regular"})")
            }
            is UINode.Stat -> {
                val weighted = if (inRow) ", weighted=true" else ""
                sb.appendLine("${prefix}Stat(label=\"${node.label}\", value=\"${node.value}\", color=${node.color ?: "default"}, size=${node.size ?: "default"}, icon=${node.icon ?: "none"}, fillsWidth=true${weighted})")
            }
            is UINode.Image -> {
                sb.appendLine("${prefix}Image(name=\"${node.systemName}\", color=${node.color ?: "default"}, size=${node.size ?: "default"})")
            }
            is UINode.Badge -> {
                sb.appendLine("${prefix}Badge(text=\"${node.text}\", color=${node.color ?: "default"})")
            }
            is UINode.Card -> {
                sb.appendLine("${prefix}Card(color=${node.color ?: "default"}, padding=${node.padding ?: 16.0}, cornerRadius=${node.cornerRadius ?: 12.0}, fillsWidth=true)")
                appendNode(sb, node.child, indent + 1, inRow = false)
            }
            is UINode.Progress -> {
                sb.appendLine("${prefix}Progress(value=${node.value}, total=${node.total}, label=${node.label ?: "none"}, color=${node.color ?: "default"})")
            }
            is UINode.Chart -> {
                sb.appendLine("${prefix}Chart(variant=${node.variant}, title=${node.title ?: "none"}) [dataPoints=${node.data.size}]")
                node.data.forEach { dp ->
                    sb.appendLine("${prefix}  DataPoint(label=\"${dp.label}\", value=${dp.value}, color=${dp.color ?: "default"})")
                }
            }
            is UINode.ListNode -> {
                sb.appendLine("${prefix}List [items=${node.items.size}]")
                node.items.forEach { item ->
                    appendNode(sb, item, indent + 1, inRow = false)
                }
            }
            is UINode.Table -> {
                sb.appendLine("${prefix}Table(title=${node.title ?: "none"}, headers=[${node.headers.joinToString(", ") { "\"$it\"" }}]) [rows=${node.rows.size}]")
                node.rows.forEach { row ->
                    sb.appendLine("${prefix}  Row([${row.joinToString(", ") { "\"$it\"" }}])")
                }
            }
            is UINode.Divider -> {
                sb.appendLine("${prefix}Divider")
            }
            is UINode.Spacer -> {
                val weighted = if (inRow) "(weighted=true)" else ""
                sb.appendLine("${prefix}Spacer${weighted}")
            }
            is UINode.Unknown -> {
                sb.appendLine("${prefix}Unknown(type=\"${node.typeName}\")")
            }
        }
    }
}

File: packages/ios/Sources/GenerativeUIDSL/RenderSnapshot.swift
import Foundation

/// Produces a canonical, deterministic text description of a UINode tree's
/// rendering behavior. Both iOS (Swift) and Android (Kotlin) generate these
/// snapshots from the same JSON fixtures — if they match, the renderers
/// produce structurally identical output.
///
/// Must produce byte-identical output to `RenderSnapshot.kt` for the same input.
public struct RenderSnapshot {

    /// Generate a snapshot for a UINode tree.
    public static func generate(_ node: UINode) -> String {
        var lines: [String] = []
        appendNode(&lines, node, indent: 0, inRow: false)
        return lines.joined(separator: "\n")
    }

    private static func appendNode(_ lines: inout [String], _ node: UINode, indent: Int, inRow: Bool) {
        let prefix = String(repeating: "  ", count: indent)

        switch node {
        case .vstack(let n):
            lines.append("\(prefix)VStack(spacing=\(n.spacing ?? 8), align=\(n.alignment ?? "center")) [children=\(n.children.count)]")
            for child in n.children {
                appendNode(&lines, child, indent: indent + 1, inRow: false)
            }

        case .hstack(let n):
            lines.append("\(prefix)HStack(spacing=\(n.spacing ?? 8), align=\(n.alignment ?? "center"), fillsWidth=true) [children=\(n.children.count)]")
            for child in n.children {
                appendNode(&lines, child, indent: indent + 1, inRow: true)
            }

        case .zstack(let n):
            lines.append("\(prefix)ZStack(align=\(n.alignment ?? "center")) [children=\(n.children.count)]")
            for child in n.children {
                appendNode(&lines, child, indent: indent + 1, inRow: false)
            }

        case .text(let n):
            lines.append("\(prefix)Text(content=\"\(n.content)\", style=\(n.style ?? "body"), color=\(n.color ?? "default"), weight=\(n.weight ?? "regular"))")

        case .stat(let n):
            let weighted = inRow ? ", weighted=true" : ""
            lines.append("\(prefix)Stat(label=\"\(n.label)\", value=\"\(n.value)\", color=\(n.color ?? "default"), size=\(n.size ?? "default"), icon=\(n.icon ?? "none"), fillsWidth=true\(weighted))")

        case .image(let n):
            lines.append("\(prefix)Image(name=\"\(n.systemName)\", color=\(n.color ?? "default"), size=\(n.size ?? "default"))")

        case .badge(let n):
            lines.append("\(prefix)Badge(text=\"\(n.text)\", color=\(n.color ?? "default"))")

        case .card(let n):
            lines.append("\(prefix)Card(color=\(n.color ?? "default"), padding=\(n.padding ?? 16), cornerRadius=\(n.cornerRadius ?? 12), fillsWidth=true)")
            appendNode(&lines, n.child, indent: indent + 1, inRow: false)

        case .progress(let n):
            lines.append("\(prefix)Progress(value=\(n.value), total=\(n.total), label=\(n.label ?? "none"), color=\(n.color ?? "default"))")

        case .chart(let n):
            lines.append("\(prefix)Chart(variant=\(n.variant), title=\(n.title ?? "none")) [dataPoints=\(n.data.count)]")
            for dp in n.data {
                lines.append("\(prefix)  DataPoint(label=\"\(dp.label)\", value=\(dp.value), color=\(dp.color ?? "default"))")
            }

        case .list(let n):
            lines.append("\(prefix)List [items=\(n.items.count)]")
            for item in n.items {
                appendNode(&lines, item, indent: indent + 1, inRow: false)
            }

        case .table(let n):
            let headers = n.headers.map { "\"\($0)\"" }.joined(separator: ", ")
            lines.append("\(prefix)Table(title=\(n.title ?? "none"), headers=[\(headers)]) [rows=\(n.rows.count)]")
            for row in n.rows {
                let cells = row.map { "\"\($0)\"" }.joined(separator: ", ")
                lines.append("\(prefix)  Row([\(cells)])")
            }

        case .divider:
            lines.append("\(prefix)Divider")

        case .spacer:
            let weighted = inRow ? "(weighted=true)" : ""
            lines.append("\(prefix)Spacer\(weighted)")
        }
    }
}
